### PR TITLE
module benchmarks initial implementation

### PR DESCRIPTION
**Pull request by jywarren**
Attempt to fix https://github.com/publiclab/image-sequencer/issues/768 - having some trouble and would love help!

### CODE REVIEW

**Comment by jywarren** [comment]
OK, this is cool!

```

    ############ Benchmarks ############
    Module average ran in: 1 milliseconds
    Module blend ran in: 1 milliseconds
    Module blur ran in: 0 milliseconds
    Module brightness ran in: 0 milliseconds
    Module channel ran in: 0 milliseconds
    Module colorbar ran in: 0 milliseconds
    Module colormap ran in: 0 milliseconds
    Module contrast ran in: 0 milliseconds
    Module convolution ran in: 1 milliseconds
    Module crop ran in: 0 milliseconds
    Module decode-qr ran in: 31 milliseconds
    Module dither ran in: 0 milliseconds
    Module draw-rectangle ran in: 0 milliseconds
    Module dynamic ran in: 1 milliseconds
    Module edge-detect ran in: 0 milliseconds
    Module fisheye-gl ran in: 1 milliseconds
    Module histogram ran in: 2 milliseconds
    Module gamma-correction ran in: 0 milliseconds
    Module gradient ran in: 0 milliseconds
    ReferenceError: $ is not defined
        at Object.setup (http://localhost:35017/bundle.js:70879:5)
        at Array.ImportImageModule (http://localhost:35017/bundle.js:70818:8)
```

So it's tripping when it gets to the `import-image` module, which is a UI module. 

https://github.com/jywarren/image-sequencer/blob/test-benchmarks/src/modules/ImportImage/Module.js

How should we skip that module?

**Comment by jywarren** [comment]
They're all still really really fast... maybe I'm doing this wrong, need to use a callback? 

```
    ############ Benchmarks ############
    Module average ran in: 0 milliseconds
    Module blend ran in: 0 milliseconds
    Module blur ran in: 0 milliseconds
    Module brightness ran in: 0 milliseconds
    Module channel ran in: 0 milliseconds
    Module colorbar ran in: 1 milliseconds
    Module colormap ran in: 0 milliseconds
    Module contrast ran in: 0 milliseconds
    Module convolution ran in: 0 milliseconds
    Module crop ran in: 1 milliseconds
    Module decode-qr ran in: 34 milliseconds
    Module dither ran in: 1 milliseconds
    Module draw-rectangle ran in: 0 milliseconds
    Module dynamic ran in: 1 milliseconds
    Module edge-detect ran in: 0 milliseconds
    Module fisheye-gl ran in: 1 milliseconds
    Module histogram ran in: 0 milliseconds
    Module gamma-correction ran in: 0 milliseconds
    Module gradient ran in: 1 milliseconds
```

File: test/core/modules/benchmark.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+test('benchmark all modules',function(t){
+
+  var start, end;
+
+  console.log("############ Benchmarks ############");
+
+  Object.keys(sequencer.modules).forEach(function forEachModule(moduleName) {
+    start = Date.now();
+    sequencer.addSteps(moduleName);
```
Comment (jywarren) [inline test/core/modules/benchmark.js]: I see - we aren't actually running the sequencer. And even when we do, we need to do something like this to measure the time:

```js
sequencer.run(function runSequencer() {
  end = Date.now();
  console.log("Module " + moduleName + " ran in: " + (end-start) + " milliseconds");
});
```

Should we look at these asynchronous iterator patterns?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of

https://stackoverflow.com/questions/4288759/asynchronous-for-cycle-in-javascript#4288992

???

Comment (jywarren) [reply test/core/modules/benchmark.js]: @publiclab/is-reviewers what do you think? Anyone have some experience with asynchronous iterators, or have another simpler way to do this? I think we have to ensure that the modules don't run at the same time or our benchmarks won't really be fair. Each step should run and be timed by itself.

Comment (jywarren) [reply test/core/modules/benchmark.js]: If anyone wants to submit a new PR based on this, that'd be great!

**Comment by HarshKhandeparkar** [comment]
@jywarren I think you have to `start = new Date.now()` inside the iterator function and then run the steps and then run `end = new Date.now()` inside the callback and the console logs also inside the callback. And I think you will have to create a new sequencer for each module since if the old sequencer has not completed yet and you run a new step, it will complete very fast.

Another way might be to use async...await. You can await the running of steps and no need to use callback.

Also shouldn't it run `sequencer.run()`?

File: test/core/modules/benchmark.js (hunk 1 above)
Comment (aashna27) [reply test/core/modules/benchmark.js]: @jywarren can I give it a shot ?

**Comment by jywarren** [comment]
Yes please do! I'd love some help!

**Comment by aashna27** [comment]
> 
> 
> @jywarren I think you have to `start = new Date.now()` inside the iterator function and then run the steps and then run `end = new Date.now()` inside the callback and the console logs also inside the callback. And I think you will have to create a new sequencer for each module since if the old sequencer has not completed yet and you run a new step, it will complete very fast.
> 
> Another way might be to use async...await. You can await the running of steps and no need to use callback.
> 
> Also shouldn't it also run `sequencer.run()` instead of only addSteps?

I like your inputs, and also promises also can be used async-await also is a great choice.

**Comment by aashna27** [comment]
> 
> Yes please do! I'd love some help!


Cool then I post here and even in the issue and submit the changes.Thanks 🛩

**Comment by jywarren** [comment]
Awesome. Feel free to open a new PR based on this branch, or to open a PR against this PR - either way! This'll be great and we can think about tracking these over time, too.

**Comment by aashna27** [comment]
> Awesome. Feel free to open a new PR based on this branch, or to open a PR against this PR - either way! This'll be great and we can think about tracking these over time, too.

Yea sounds great !!! 😄

**Comment by jywarren** [comment]
> How should we skip that module? (`image-input`)

I was thinking on this and i think we should set a property of that module like `module.type = 'input'` perhaps. Then, as we resolve why other properties are being added to `sequencer.modules` and fix that, we should have just a list of basic standard modules in `sequencer.modules`.

**Comment by tech4GT** [comment]
@jywarren The run method you mentioned is the best way to do this! We need to add a module into sequencer then run the sequencer. We can do this for each module separately.

**Comment by tech4GT** [comment]
I think I fixed this, pushing in a while :)

**Comment by tech4GT** [comment]
moving this to #781

**Pull request closed without merging**